### PR TITLE
More sec barricade changes/general cover changes

### LINKED ISSUE #12537: Wooden barricades have no hit delay
Can hit/destroy them extremely quickly

### PR DESCRIPTION
**Pull request by imtakingabreakdontatme**
Sec barricades and wooden barricades have been unified under one object type (barricade). Barricades have a chance of blocking gunfire, but you can fire past them if you are standing adjacent.

Sec barricades:

-Are now deployed via grenade. The grenade has three modes (single barrier, three vertical, three horizontal). The barriers take a few seconds to fully deploy once spawned by the grenade, and they can be moved during this time (mostly for balance reasons).

-Are no longer able to be moved once deployed (being able to emag/sec ID past them made them worthless)

-Have more health (but dont take half damage from melee anymore)

Cover: 

-Repaired with a welding tool

-Melee delay has been fixed

Fixes https://github.com/tgstation/-tg-station/issues/12537

:cl: Kor
rscadd: Security barricades are now deployed via grenade. You can no longer move them after they've been deployed, you must destroy them.
rscadd: You can now shoot through wooden barricades while adjacent, while any other gunfire has a 50% chance of being blocked.
/:cl:

### CODE REVIEW

**Comment by Supermichael777** [comment]
Deploying level 2 Baricade

File: code/game/machinery/deployable.dm [hunk 1]
```diff
@@ -254,24 +77,76 @@ for reference:
...
+
+/obj/item/weapon/grenade/barrier
+	name = "barrier grenade"
+	desc = "Instant cover. Alt+click to toggle modes."
+	icon = 'icons/obj/grenade.dmi'
+	icon_state = "flashbang"
+	item_state = "flashbang"
+	var/mode = "single"
+
+/obj/item/weapon/grenade/barrier/AltClick()
```
Comment (RemieRichards) [inline code/game/machinery/deployable.dm]: AltClick has a user argument you poo.

**Comment by Jalleo** [comment]
Maybe make a way which can pack up the barrier again or make the grenade a projectile and a gun that you can reload the barrier into.

File: code/game/machinery/deployable.dm [hunk 2]
```diff
@@ -254,24 +77,76 @@ for reference:
...
+		if("single")
+			mode = "vertical"
+		if("vertical")
+			mode = "horizantal"
+		if("horizantal")
+			mode = "single"
+
+	if(ismob(loc))
+		var/mob/M = loc
+		M << "[src] is now in [mode] mode."
```
Comment (RemieRichards) [inline code/game/machinery/deployable.dm]: seeeee aboooooove

File: code/game/machinery/deployable.dm [hunk 3]
```diff
@@ -254,24 +77,76 @@ for reference:
...
+	..()
+	switch(mode)
+		if("single")
+			mode = "vertical"
+		if("vertical")
+			mode = "horizantal"
+		if("horizantal")
+			mode = "single"
+
+	if(ismob(loc))
```
Comment (RemieRichards) [inline code/game/machinery/deployable.dm]: seeee abooooove

File: code/game/machinery/deployable.dm [hunk 4]
```diff
@@ -254,24 +77,76 @@ for reference:
...
+		visible_message("<span class='warning'>[src] deploys!</span>")
+
+
+/obj/item/weapon/grenade/barrier
+	name = "barrier grenade"
+	desc = "Instant cover. Alt+click to toggle modes."
+	icon = 'icons/obj/grenade.dmi'
+	icon_state = "flashbang"
+	item_state = "flashbang"
+	var/mode = "single"
```
Comment (RemieRichards) [inline code/game/machinery/deployable.dm]: defines pls

**Comment by Bawhoppen** [comment]
also can I just say I really I've always hated the sprite of the wooden barricade

**Comment by Zilenan91** [comment]
We could always make these an armory item like holotape projectors that can be used to vend barricades rather than making them a grenade.

**Comment by KorPhaeron** [comment]
How can I fire it out of a grenade launcher if it isnt a grenade?

**Comment by Zilenan91** [comment]
Imagination my good friend

**Comment by Dorsisdwarf** [comment]
Add a clusterbang version.

File: code/game/machinery/deployable.dm [hunk 5]
```diff
@@ -1,81 +1,34 @@
...
+/obj/structure/barricade/proc/take_damage(damage, leave_debris=1, message)
 	health -= damage
 	if(health <= 0)
 		if(message)
 			visible_message(message)
 		else
-			visible_message("<span class='warning'>The barricade is smashed apart!</span>")
-		if(leave_debris)
-			new /obj/item/stack/sheet/mineral/wood(get_turf(src), 3)
+			visible_message("<span class='warning'>The [src] is smashed apart!</span>")
```
Comment (duncathan) [inline code/game/machinery/deployable.dm]: `\The [src]`

File: code/game/machinery/deployable.dm [hunk 6]
```diff
@@ -84,168 +37,42 @@ for reference:
...
 		..()
-		var/damage = 0
-		switch(P.damage_type)
-			if(BURN)
-				damage = P.damage * 1.5
-			if(BRUTE)
-				damage = P.damage * 1
-		take_damage(damage)
+		take_damage(P.damage*ranged_damage_modifier)
+		visible_message("<span class='warning'>The [src] is hit by [P]!</span>")
```
Comment (duncathan) [inline code/game/machinery/deployable.dm]: `\The [src]`

File: code/game/machinery/deployable.dm [hunk 7]
```diff
@@ -84,168 +37,42 @@ for reference:
...
 
-/obj/structure/barricade/wooden/ex_act(severity, target)
+/obj/structure/barricade/ex_act(severity, target)
 	switch(severity)
 		if(1)
-			visible_message("<span class='warning'>The barricade is blown apart!</span>")
+			visible_message("<span class='warning'>The [src] is blown apart!</span>")
```
Comment (duncathan) [inline code/game/machinery/deployable.dm]: See above

File: code/game/machinery/deployable.dm [hunk 8]
```diff
@@ -84,168 +37,42 @@ for reference:
...
-		visible_message("<span class='danger'>BZZzZZzZZzZT</span>")
-		return
-
-/obj/machinery/deployable/barrier/ex_act(severity)
-	switch(severity)
-		if(1)
-			explode()
-		if(2)
-			take_damage(25)
+			take_damage(25, message = "<span class='warning'>The [src] is blown apart!</span>")
```
Comment (duncathan) [inline code/game/machinery/deployable.dm]: Just fix every one of these OK

File: code/game/machinery/deployable.dm (hunk 5 above)
Comment (tkdrg) [inline code/game/machinery/deployable.dm]: `\The [src]`

File: code/game/machinery/deployable.dm [hunk 9]
```diff
@@ -84,168 +37,42 @@ for reference:
...
-		..()
-		var/damage = 0
-		switch(W.damtype)
-			if(BURN)
-				damage = W.force * 1.5
-			if(BRUTE)
-				damage = W.force * 1
-		take_damage(damage)
+		user.changeNext_move(CLICK_CD_MELEE)
+		visible_message("<span class='warning'>[user] hits [src] with [I]!</span>", "<span class='warning'>You hit [src] with [I]!</span>")
```
Comment (tkdrg) [inline code/game/machinery/deployable.dm]: `\the [src]` on both

File: code/game/machinery/deployable.dm (hunk 6 above)
Comment (tkdrg) [inline code/game/machinery/deployable.dm]: `\The`

File: code/game/machinery/deployable.dm (hunk 7 above)
Comment (tkdrg) [inline code/game/machinery/deployable.dm]: `\The`

File: code/game/machinery/deployable.dm [hunk 10]
```diff
@@ -254,24 +81,79 @@ for reference:
...
-	explosion(src.loc,-1,-1,0)
-	if(src)
-		qdel(src)
+/////BARRICADE TYPES///////
+
+/obj/structure/barricade/wooden
+	name = "wooden barricade"
+	desc = "This space is blocked off by a wooden barricade."
+	icon = 'icons/obj/structures.dmi'
+	icon_state = "woodenbarricade"
```
Comment (tkdrg) [inline code/game/machinery/deployable.dm]: So I can fix wooden barricades with a welding fuel? wtf

File: code/game/machinery/deployable.dm [hunk 11]
```diff
@@ -254,24 +81,79 @@ for reference:
...
+	..()
+	switch(mode)
+		if(SINGLE)
+			mode = VERTICAL
+		if(VERTICAL)
+			mode = HORIZONTAL
+		if(HORIZONTAL)
+			mode = SINGLE
+
+	user << "[src] is now in [mode] mode."
```
Comment (tkdrg) [inline code/game/machinery/deployable.dm]: Could add an action button for this as well, just for completeness and intuitiveness.

File: code/game/machinery/deployable.dm [hunk 12]
```diff
@@ -254,24 +81,79 @@ for reference:
...
+
+/obj/item/weapon/grenade/barrier/prime()
+	new /obj/structure/barricade/security(get_turf(src.loc))
+	switch(mode)
+		if(VERTICAL)
+			new /obj/structure/barricade/security(get_step(src, NORTH))
+			new /obj/structure/barricade/security(get_step(src, SOUTH))
+		if(HORIZONTAL)
+			new /obj/structure/barricade/security(get_step(src, EAST))
+			new /obj/structure/barricade/security(get_step(src, WEST))
```
Comment (tkdrg) [inline code/game/machinery/deployable.dm]: Should probably check if these tiles are valid, so we don't end up with barricades inside walls or inside windows.

File: code/game/machinery/deployable.dm (hunk 10 above)
Comment (KorPhaeron) [inline code/game/machinery/deployable.dm]: The attack code/construction/deconstruction stuff is so awful I'm not sure how else to generalize it

File: code/game/machinery/deployable.dm (hunk 12 above)
Comment (KorPhaeron) [inline code/game/machinery/deployable.dm]: How do I check if a turf is completely clear of dense objects? Must be a proc somewhere for it

File: code/game/machinery/deployable.dm (hunk 11 above)
Comment (KorPhaeron) [inline code/game/machinery/deployable.dm]: Correct me if I'm wrong but action buttons generally default to attack_hand, which in this case arms the grenade

Comment (tkdrg) [inline code/game/machinery/deployable.dm]: Action buttons default to calling `ui_action_click()`, which defaults to `attack_self()`. Just make this AltClick stuff a proc, and call it from ui_action_click() and AltClick(). Then all you should have to do is set an appropriate action_button_name.

File: code/game/machinery/deployable.dm (hunk 10 above)
Comment (tkdrg) [inline code/game/machinery/deployable.dm]: just replace the welding stuff in attackby() with a call to a being_welded() or something like that, and override being_welded() to just return on the wooden child

File: code/game/machinery/deployable.dm (hunk 12 above)
Comment (tkdrg) [inline code/game/machinery/deployable.dm]: just do an `if(is_blocked_turf(target_turf))`

**Comment by Lumipharon** [comment]
Having a old version of the sec barrier (that can be ID'd and moved) still around would be handy.
Sure, emags can get past them, but people without emags or sec ID's (ie: most antags) can't, so they're still useful.

**Pull request merged**